app/src/main/kotlin/com/github/gotify/messages/MessagesActivity.kt

```diff
@@ -171,52 +201,56 @@ internal class MessagesActivity :
 
     private fun refreshAll() {
         CoilInstance.evict(this)
-        startActivity(Intent(this, InitializationActivity::class.java))
-        finish()
+        messagesViewModel.onRefreshAll()
     }
 
     private fun onRefresh() {
         CoilInstance.evict(this)
-        viewModel.messages.clear()
-        launchCoroutine {
-            loadMore(viewModel.appId).forEachIndexed { index, message ->
-                if (message.image != null) {
-                    listMessageAdapter.notifyItemChanged(index)
-                }
-            }
-        }
+        messagesViewModel.onRefreshMessages()
     }
 
     private fun openDocumentation() {
         val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse("https://gotify.net/docs/pushmsg"))
         startActivity(browserIntent)
     }
 
-    private fun onUpdateApps(applications: List<Application>) {
+    private fun onUpdateApps(applicationsState: List<ApplicationState>) {
         val menu: Menu = binding.navView.menu
         menu.removeGroup(R.id.apps)
-        viewModel.targetReferences.clear()
-        updateMessagesAndStopLoading(viewModel.messages[viewModel.appId])
-        var selectedItem = menu.findItem(R.id.nav_all_messages)
-        applications.indices.forEach { index ->
-            val app = applications[index]
-            val item = menu.add(R.id.apps, index, APPLICATION_ORDER, app.name)
-            item.isCheckable = true
-            if (app.id == viewModel.appId) selectedItem = item
-            val t = Utils.toDrawable { icon -> item.icon = icon }
-            viewModel.targetReferences.add(t)
-            val request = ImageRequest.Builder(this)
-                .data(Utils.resolveAbsoluteUrl(viewModel.settings.url + "/", app.image))
-                .error(R.drawable.ic_alarm)
-                .placeholder(R.drawable.ic_placeholder)
-                .size(100, 100)
-                .target(t)
-                .build()
-            CoilInstance.get(this).enqueue(request)
+        applicationsState.forEachIndexed { index, applicationState ->
+            menu.add(
+                R.id.apps,
+                index,
+                APPLICATION_ORDER,
+                applicationState.application.name
+            ).apply {
+                setOnMenuItemClickListener {
+                    messagesViewModel.onSelectApplication(applicationState.application)
+                    binding.drawerLayout.closeDrawer(GravityCompat.START)
+                    true
+                }
+                if (applicationState.unreadCount > 0) {
+                    setActionView(R.layout.action_menu_counter)
+                    actionView?.findViewById<TextView>(
+                        R.id.counter
+                    )?.text = applicationState.counterLabel
```

Comment:
Well, It is storing messages there, but this list is not populated when the app starts, then, if the user doesn't navigate to every app section, the list of messages is not updated.
Apart of that, the data within `MessageStateHolder` is not "announced" outside of the instance, and the `ViewModel` doesn't act as a ViewModel but as a "data provider" that you need to ask for data.

Reply:
> Well, It is storing messages there, but this list is not populated when the app starts, then, if the user doesn't navigate to every app section, the list of messages is not updated.

I get the last part, but the first part seems a little dishonest. The current implementation doesn't populate all messages from all apps because it doesn't need to. It sounds like "It's not possible with the current implementation, so there must be a rewrite", which I don't buy, changing the current implementation to load all messages from all apps should be simpler than wholly rewriting the message handling.

(not saying the current implementation should be adjusted like this, it only about the argumentation)